src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -190,5 +195,205 @@ public static bool AreOnSameLine(this SourceText text, SyntaxToken token1, Synta
                 token2.RawKind != 0 &&
                 text.Lines.IndexOf(token1.Span.End) == text.Lines.IndexOf(token2.SpanStart);
         }
+
+        // 32KB. comes from SourceText char buffer size and less than large object size
+        internal const int SourceTextLengthThreshold = 32 * 1024 / sizeof(char);
+
+        public static void WriteTo(this SourceText sourceText, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            // Source length
+            var length = sourceText.Length;
+            writer.WriteInt32(length);
+
+            // if source is small, no point on optimizing. just write out string
+            if (length < SourceTextLengthThreshold)
+            {
+                writer.WriteString(sourceText.ToString());
+            }
+            else
+            {
+                // if bigger, write out as chunks
+                WriteChunksTo(sourceText, writer, length, cancellationToken);
+            }
+        }
+
+        private static void WriteChunksTo(SourceText sourceText, ObjectWriter writer, int length, CancellationToken cancellationToken)
+        {
+            // chunk size
+            var buffer = SharedPools.CharArray.Allocate();
+            writer.WriteInt32(buffer.Length);
+
+            // number of chunks
+            var numberOfChunks = 1 + (length / buffer.Length);
+            writer.WriteInt32(numberOfChunks);
+
+            // write whole chunks
+            try
+            {
+                var offset = 0;
+                for (var i = 0; i < numberOfChunks; i++)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var count = Math.Min(buffer.Length, length - offset);
```

Comment:
due to this condition, "if ((i < numberOfChunks - 1) || (count == buffer.Length))", I think having just one loop is better.